### ISSUE TITLE
support RTOS family in the plan?

### ISSUE THREAD

**Issue by yorkie** [CraneStation/wasmtime#420]
The RTOS might be on most low-end device, that would be great if WASI is able to land on those. 

It's related with https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#proc_exit and the multithreading discussion #419, most RTOS platforms do not have child process or threading, we should be careful to include the system **interfaces** if we want to land those :)

**Comment by tschneidereit** [comment]
Wasmtime will definitely be an option for these environments, yes. Note that any and all WASI modules will be optional and you'll be able to configure Wasmtime to not include them.

**Comment by kubkon** [comment]
Heck yeah, RTOS support would be amazing. +10 here! ;-)

**Comment by yorkie** [comment]
RTOS will not have the process or threading, so we have to flag the related
API or impl to user-land, do u guys mind that i create a pr that removes
process?

:p

**Comment by tschneidereit** [comment]
Can you explain a bit more what you're proposing to remove exactly? As I said, we'll definitely have the ability to build Wasmtime without all optional features. Multithreading is definitely one of those. And as far as I'm aware, we don't currently have any child processes, so there's nothing to remove.

**Comment by yorkie** [comment]
What I'm proposed that is:

- https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#__wasi_proc_exit
- https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#__wasi_proc_raise

It might be: `__wasi_exit()` and `__wasi_raise()`, and change the "process" to "execution", too. The concept "process" is different on these platforms like: linux, unix and RTOS family.

For example, at FreeRTOS, it only owns itself "**execution unit**" namely [task](https://www.freertos.org/RTOS-task-states.html) to replace the process or thread, so what's I'm proposed is to abstract process/task/thread in the wasi layer.

**Comment by yorkie** [comment]
> Can you explain a bit more what you're proposing to remove exactly? As I said, we'll definitely have the ability to build Wasmtime without all optional features. Multithreading is definitely one of those. And as far as I'm aware, we don't currently have any child processes, so there's nothing to remove.

Optional features answered my question absolutely, the rest is doing port/re-implement the wastime on RTOS, thanks :)

**Comment by kubkon** [comment]
> What I'm proposed that is:
> 
> * https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#__wasi_proc_exit
> * https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#__wasi_proc_raise
> 
> It might be: `__wasi_exit()` and `__wasi_raise()`, and change the "process" to "execution", too. The concept "process" is different on these platforms like: linux, unix and RTOS family.
> 
> For example, at FreeRTOS, it only owns itself "**execution unit**" namely [task](https://www.freertos.org/RTOS-task-states.html) to replace the process or thread, so what's I'm proposed is to abstract process/task/thread in the wasi layer.

A small clarification here. `__wasi_proc_raise()` is actually meant to be dropped entirely from WASI and hence, we don’t even implement it in [wasi-common]. As far as `__wasi_proc_exit()` is concerned, it indeed currently uses `std::process::exit()` to terminate itself by terminating the process it is running in. I guess, for RTOS, the tricky bit would be to find the equivalent in Rust, but I might be talking crap here so please correct me if I’m wrong. Also, cc @sunfishcode as it might be an interest to you given that we’ve already discussed a little about this and similar use cases offline in the past.

[wasi-common]: https://github.com/CraneStation/wasi-common

**Comment by yorkie** [comment]
I googled something related and find this: https://docs.rs/freertos_rs/0.3.0/freertos_rs/, it might be a quick start, however we have to work on abstracting others, too.

**Comment by sunfishcode** [comment]
I agree that `proc_raise` doesn't add much value in WASI in its current form and should be removed. 

For `proc_exit`, I expect that when wasm gets [exception handling](https://github.com/WebAssembly/exception-handling/), `proc_exit` will go away entirely, and we'll instead just use a `throw`, throwing a non-HLL-catchable exception all the way out to the code that calls the user's `main` function.

**Comment by fitzgen** [comment]
Wasmtime supports no-std, WASI is defined elsewhere; issue is no longer relevant.

**Comment by ricochet** [comment]
See also https://docs.wasmtime.dev/stability-tiers.html#tier-3
